### PR TITLE
Soft components support

### LINKED ISSUE #1431: Add support for conditional components using `@component`
Now you can only specify what plugins your theme depends on. It doesn't make sense for theme authors to bake in support for every popular plugin if it pulls in all those dependencies as well. However, a theme might have support for it but not require it. This would need a mechanism for themes to check if a plugin is activated and then include code.

This would also be a huge benefit to the theme marketplace since the users can search for themes that supports RainLab.Blog, for example.

### PR DESCRIPTION
**Pull request by viamage**
This is a proposal to resolve #1431, making it a PR per LukeTowers request.

### CODE REVIEW

**Comment by viamage** [comment]
So first thing to discuss here would be if try-catch approach is best here.

Second would be logging - I'm quite sure we should inform website admin that there is something wrong in theme (and lacking component that is called in theme is something wrong in theme, right?) - but if we log it or not depends on what this enhancement use cases are. I personally would keep it, but I understand reasons to possibly remove it.

**Comment by LukeTowers** [comment]
Soft components don't need to be logged when they're missing, they're explicitly defined as optional so them not being present isn't a problem. Could you make a PR to the docs to document this feature and remove the logging?

**Comment by Samuell1** [comment]
@viamage Hey, any news on this?

File: modules/cms/classes/Controller.php [hunk 1]
```diff
@@ -1237,31 +1237,55 @@ public function param($name, $default = null)
...
         $manager = ComponentManager::instance();
 
-        if ($addToLayout) {
-            if (!$componentObj = $manager->makeComponent($name, $this->layoutObj, $properties)) {
-                throw new CmsException(Lang::get('cms::lang.component.not_found', ['name'=>$name]));
-            }
-
-            $componentObj->alias = $alias;
-            $this->vars[$alias] = $this->layout->components[$alias] = $componentObj;
+        if($isSoftComponent = $this->isSoftComponent($name)){
```
Comment (LukeTowers) [inline modules/cms/classes/Controller.php]: space after if please `if (`

File: modules/cms/classes/Controller.php [hunk 2]
```diff
@@ -1237,31 +1237,55 @@ public function param($name, $default = null)
...
+            $this->setComponentPropertiesFromParams($componentObj);
+            $componentObj->init();
+
+            return $componentObj;
+            /**
+             * We catch two exceptions below, as missing plugin can throw both and catch \Exception would be too global.
+             * We also debug-log it, to not be TOO silent about this.
+             */
+        } catch(CmsException $e){
+            \Log::debug('Missing component! '. $e->getMessage() . ' '. $e->getTraceAsString());
```
Comment (LukeTowers) [inline modules/cms/classes/Controller.php]: remove log please

File: modules/cms/classes/Controller.php [hunk 3]
```diff
@@ -1237,31 +1237,55 @@ public function param($name, $default = null)
...
-            $this->vars[$alias] = $this->page->components[$alias] = $componentObj;
+            $this->setComponentPropertiesFromParams($componentObj);
+            $componentObj->init();
+
+            return $componentObj;
+            /**
+             * We catch two exceptions below, as missing plugin can throw both and catch \Exception would be too global.
+             * We also debug-log it, to not be TOO silent about this.
+             */
+        } catch(CmsException $e){
```
Comment (LukeTowers) [inline modules/cms/classes/Controller.php]: space after `catch` please (`catch (`)

File: modules/cms/classes/Controller.php [hunk 4]
```diff
@@ -1237,31 +1237,55 @@ public function param($name, $default = null)
...
+            \Log::debug('Missing component! '. $e->getMessage() . ' '. $e->getTraceAsString());
+            if(!$isSoftComponent){
```
Comment (LukeTowers) [inline modules/cms/classes/Controller.php]: space after if please `if (`

File: modules/cms/classes/Controller.php [hunk 5]
```diff
@@ -1237,31 +1237,55 @@ public function param($name, $default = null)
...
+                throw $e;
+            }
+        } catch(SystemException $e){
```
Comment (LukeTowers) [inline modules/cms/classes/Controller.php]: space after catch please (`catch (`)

File: modules/cms/classes/Controller.php [hunk 6]
```diff
@@ -1237,31 +1237,55 @@ public function param($name, $default = null)
...
+            \Log::debug('Missing component! '. $e->getMessage() . ' '. $e->getTraceAsString());
```
Comment (LukeTowers) [inline modules/cms/classes/Controller.php]: remove log please

File: modules/cms/classes/Controller.php [hunk 7]
```diff
@@ -1237,31 +1237,55 @@ public function param($name, $default = null)
...
+            if(!$isSoftComponent){
```
Comment (LukeTowers) [inline modules/cms/classes/Controller.php]: space after if please `if (`

File: modules/cms/classes/Controller.php [hunk 8]
```diff
@@ -1411,4 +1435,25 @@ protected function verifyCsrfToken()
             $token
         );
     }
+
+    /**
+     * Removes '@' from soft component name
+     * @param string $label
+     * @return string
+     */
+    private function parseComponentLabel($label){
```
Comment (LukeTowers) [inline modules/cms/classes/Controller.php]: protected instead of private please

File: modules/cms/classes/Controller.php [hunk 9]
```diff
@@ -1411,4 +1435,25 @@ protected function verifyCsrfToken()
...
+        if($this->isSoftComponent($label)){
```
Comment (LukeTowers) [inline modules/cms/classes/Controller.php]: space after if

File: modules/cms/classes/Controller.php [hunk 10]
```diff
@@ -1411,4 +1435,25 @@ protected function verifyCsrfToken()
...
+            return str_replace('@','',$label);
```
Comment (LukeTowers) [inline modules/cms/classes/Controller.php]: space after arguments (`'@', '', $label`)

File: modules/cms/classes/Controller.php [hunk 11]
```diff
@@ -1411,4 +1435,25 @@ protected function verifyCsrfToken()
...
+        }
+        return $label;
+    }
+
+    /**
+     * Checks if component name has @.
+     * @param string $label
+     * @return bool
+     */
+    private function isSoftComponent($label){
```
Comment (LukeTowers) [inline modules/cms/classes/Controller.php]: protected instead of private please

**Comment by DanHarrin** [comment]
Want these changes completed as a new PR?

**Comment by w20k** [comment]
@DanHarrin would be nice! 👍 You could add credits to @viamage.

**Comment by w20k** [comment]
Closing this PR in favor of #4539

**Pull request closed without merging**